orderer/common/server/main.go

```diff
@@ -113,58 +112,91 @@ func Main() {
-				logger.Panicf("Failed validating bootstrap block: %v", err)
-			}
-			sysChanLastConfigBlock := extractSysChanLastConfig(lf, bootstrapBlock)
-			clusterBootBlock = selectClusterBootBlock(bootstrapBlock, sysChanLastConfigBlock)
-
-			typ := consensusType(bootstrapBlock, cryptoProvider)
-			clusterType = isClusterType(clusterBootBlock, cryptoProvider)
-			if clusterType {
-				logger.Infof("Setting up cluster for orderer type %s", typ)
-				clusterClientConfig = initializeClusterClientConfig(conf)
-				clusterDialer = &cluster.PredicateDialer{
-					Config: clusterClientConfig,
-				}
 
-				r = createReplicator(lf, bootstrapBlock, conf, clusterClientConfig.SecOpts, signer, cryptoProvider)
-				// Only clusters that are equipped with a recent config block can replicate.
-				if conf.General.BootstrapMethod == "file" {
-					r.replicateIfNeeded(bootstrapBlock)
-				}
+	// This happens only when the orderer is operating with a system channel.
+	if clusterBootBlock != nil {
+		consensusTypeName := consensusType(clusterBootBlock, cryptoProvider)
+		_, clusterType = clusterTypes[consensusTypeName]
 
-				if reuseGrpcListener = reuseListener(conf, typ); !reuseGrpcListener {
-					clusterServerConfig, clusterGRPCServer = configureClusterListener(conf, serverConfig, ioutil.ReadFile)
-				}
+		if clusterType {
+			logger.Infof("Setting up cluster for orderer type %s", consensusTypeName)
+			clusterClientConfig = initializeClusterClientConfig(conf)
+			clusterDialer = &cluster.PredicateDialer{
+				Config: clusterClientConfig,
+			}
 
-				// If we have a separate gRPC server for the cluster,
-				// we need to update its TLS CA certificate pool.
-				serversToUpdate = append(serversToUpdate, clusterGRPCServer)
+			if reuseGrpcListener = reuseListener(conf); !reuseGrpcListener {
+				clusterServerConfig, clusterGRPCServer = configureClusterListener(conf, serverConfig, ioutil.ReadFile)
 			}
-			// Are we bootstrapping?
-			if len(lf.ChannelIDs()) == 0 {
-				initializeBootstrapChannel(clusterBootBlock, lf)
-			} else {
-				logger.Info("Not bootstrapping because of existing channels")
+
+			// If we have a separate gRPC server for the cluster,
+			// we need to update its TLS CA certificate pool.
+			serversToUpdate = append(serversToUpdate, clusterGRPCServer)
+
+			// Are we bootstrapping with a block with number >0 ?
+			// Only clusters that are equipped with a recent config block (number i.e. >0) can replicate.
+			// This will replicate all channels if the clusterBootBlock number > system-channel height (i.e. there is a gap in the ledger).
+			repInitiator = onboarding.NewReplicationInitiator(lf, clusterBootBlock, conf, clusterClientConfig.SecOpts, signer, cryptoProvider)
+			if conf.General.BootstrapMethod == "file" {
+				repInitiator.ReplicateIfNeeded(clusterBootBlock)
 			}
 		}
+	} else {
+		logger.Infof("Setting up cluster without a system channel")
+		clusterType = true
+		clusterClientConfig = initializeClusterClientConfig(conf)
```

Comment:
Since lines 188 thru 199 actually repition of lines 164-175, I think this needs to be extracted into funciton.

Reply:
done, removed the duplicate code